executor/window.go

```diff
@@ -0,0 +1,160 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/executor/windowfuncs"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+// WindowExec is the executor for window functions.
+type WindowExec struct {
+	baseExecutor
+
+	groupChecker *groupChecker
+	inputIter    *chunk.Iterator4Chunk
+	inputRow     chunk.Row
+	groupRows    []chunk.Row
+	childResults []*chunk.Chunk
+	windowFunc   windowfuncs.WindowFunc
+	executed     bool
+	childCols    []*expression.Column
+}
+
+// Close implements the Executor Close interface.
+func (e *WindowExec) Close() error {
+	e.childResults = nil
+	return errors.Trace(e.baseExecutor.Close())
+}
+
+// Next implements the Executor Next interface.
+func (e *WindowExec) Next(ctx context.Context, chk *chunk.Chunk) error {
+	if span := opentracing.SpanFromContext(ctx); span != nil && span.Tracer() != nil {
+		span1 := span.Tracer().StartSpan("windowExec.Next", opentracing.ChildOf(span.Context()))
+		defer span1.Finish()
+	}
+	if e.runtimeStats != nil {
+		start := time.Now()
+		defer func() { e.runtimeStats.Record(time.Now().Sub(start), chk.NumRows()) }()
+	}
+	chk.Reset()
+	if e.windowFunc.HasRemainingResults() {
+		err := e.appendResult2Chunk(chk)
+		if err != nil {
+			return err
+		}
+	}
+	for !e.executed && (chk.NumRows() == 0 || chk.RemainedRows(chk.NumCols()-1) > 0) {
+		err := e.consumeOneGroup(ctx, chk)
+		if err != nil {
+			e.executed = true
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (e *WindowExec) consumeOneGroup(ctx context.Context, chk *chunk.Chunk) error {
+	for !e.executed {
+		if err := e.fetchChildIfNecessary(ctx, chk); err != nil {
+			return errors.Trace(err)
+		}
+		for ; e.inputRow != e.inputIter.End(); e.inputRow = e.inputIter.Next() {
+			meetNewGroup, err := e.groupChecker.meetNewGroup(e.inputRow)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if meetNewGroup {
+				err := e.consumeGroupRows(chk)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				err = e.appendResult2Chunk(chk)
+				if err != nil {
+					return errors.Trace(err)
+				}
+			}
+			e.groupRows = append(e.groupRows, e.inputRow)
+			if meetNewGroup {
+				e.inputRow = e.inputIter.Next()
+				return nil
+			}
+		}
+	}
+	return nil
+}
+
+func (e *WindowExec) consumeGroupRows(chk *chunk.Chunk) error {
+	if len(e.groupRows) == 0 {
+		return nil
+	}
+	e.copyChk(chk)
```

Comment:
This seems not right when a group is larger than a chunk, we would only call `ExhaustResult` for the last chunk in the group?

Reply:
In https://github.com/pingcap/tidb/pull/8899/files#diff-509643ad918102b310cd4be4c55fc644R58, we would first check if there are any remaining results, so I think it is correct.

Reply:
OK, copyChk would check `chk.NumRows() > 0` before removing a chunk from `e.childResults`, so it is right. Then there is another question: `ProcessOneChunk` would be called multiple times for a single `dest` chunk if a group is larger than a chunk, then how do we generate `row_number()` for rows in `dest` in `ProcessOneChunk`?

Reply:
Why would `ProcessOneChunk` be called multiple times for a single `dest` chunk if a group is larger than a chunk? I think it would be only called once for a `dest` chunk.

Reply:
If a group is larger than a chunk, `fetchChildIfNecessary` would be called multiple times in `consumeOneGroup`? then `fetchChildIfNecessary` would call `consumeGroupRows -> e.copyChk/ProcessOneChunk`, for the latter calls of `fetchChildIfNecessary`, in `copyChk`, `chk` would not be copied because `chk.NumRows() > 0` is satisfied now, then this `chk` would be passed to `ProcessOneChunk` multiple times?

Reply:
Oh, you are right, we need to quit after processing one chunk.